modules/utxo-lib/src/bitgo/zcash/ZcashPsbt.ts

```diff
@@ -22,6 +32,68 @@ export class ZcashPsbt extends UtxoPsbt<ZcashTransaction<bigint>> {
     );
   }
 
+  /**
+   * In version < 5 of Zcash transactions, the consensus branch ID is not serialized in the transaction
+   * whereas in version 5 it is. If the transaction is less than a version 5, set the consensus branch id
+   * in the global map in the psbt. If it is a version 5 transaction, throw an error if the consensus
+   * branch id is set in the psbt (because it should be on the transaction already).
+   * @param buffer Psbt buffer
+   * @param opts options
+   */
+  static fromBuffer(buffer: Buffer, opts: PsbtOpts): UtxoPsbt<UtxoTransaction<bigint>> {
+    const psbt = super.fromBuffer(buffer, opts) as ZcashPsbt;
+
+    // Read `consensusBranchId` from the global-map
+    let consensusBranchId: number | undefined = undefined;
+    psbt.data.globalMap.unknownKeyVals?.forEach(({ key, value }, i) => {
+      if (key.equals(CONSENSUS_BRANCH_ID_KEY)) {
+        consensusBranchId = value.readUint32BE();
+      }
+    });
+    switch (psbt.tx.version) {
+      case 4:
+      case ZcashTransaction.VERSION4_BRANCH_CANOPY:
+      case ZcashTransaction.VERSION4_BRANCH_NU5:
+        if (!consensusBranchId || !psbt.data.globalMap.unknownKeyVals) {
+          throw new Error('Could not find consensus branch id on psbt for version 4 Zcash transaction');
+        }
+        psbt.tx.consensusBranchId = consensusBranchId;
+        psbt.data.globalMap.unknownKeyVals = psbt.data.globalMap.unknownKeyVals.filter(
+          ({ key }) => key !== CONSENSUS_BRANCH_ID_KEY
+        );
+        return psbt;
+      case 5:
+      case ZcashTransaction.VERSION5_BRANCH_NU5:
+        if (consensusBranchId) {
+          throw new Error('Found consensus branch id in psbt global-map for version 5 Zcash transaction');
+        }
+        return psbt;
+      default:
+        throw new Error(`Unsupported transaction version ${psbt.tx.version}`);
+    }
+  }
+
+  /**
+   * If it is a version 4 transaction, add the consensus branch id to
+   * the global map. If it is a version 5 transaction, just return the
+   * buffer because the consensus branch id is already serialized in
+   * the transaction.
+   */
+  toBuffer(): Buffer {
+    if (this.tx.version === 5 || this.tx.version === ZcashTransaction.VERSION5_BRANCH_NU5) {
+      return super.toBuffer();
+    }
+    const value = Buffer.alloc(4);
+    value.writeUint32BE(this.tx.consensusBranchId);
+    this.addUnknownKeyValToGlobal({ key: CONSENSUS_BRANCH_ID_KEY, value });
+    if (!this.data.globalMap.unknownKeyVals) {
+      throw new Error('Failed adding consensus branch id to unknownKeyVals');
+    }
+    const buff = super.toBuffer();
+    this.data.globalMap.unknownKeyVals.pop();
```

Comment:
Not a blocker, but I have some dumb questions:

why do we pop the `consensusBranchId` from the `unknownKeyVals` here? Is this so we make it look like `toBuffer()` is side effect free?

Also, I'm curious why we don't add the consensus branch id to the `unknownKeyVals` upon construction (or at least, once we know it's a version 5 tx)?

Reply:
> why do we pop the `consensusBranchId` from the `unknownKeyVals` here? Is this so we make it look like `toBuffer()` is side effect free?

Yes

> Also, I'm curious why we don't add the consensus branch id to the `unknownKeyVals` upon construction

This was the discussion that brandon and I had today because I originally had it this way. The problem with doing it this way is that the property of `consensusBranchId` lives on the transaction object, and not on the PSBT (this is a similar behavior as was seen in the [ZcashTransactionBuilder](https://github.com/BitGo/BitGoJS/blob/master/modules/utxo-lib/src/bitgo/zcash/ZcashTransactionBuilder.ts#L105). If we add the consensus branch id to the `unknownKeyVals` map, it would mean that we would have to manage (potentially) two separate states (one on the transaction object and the other in the `unknownKeyVals`). By putting this logic in the serialization phase, we only have to manage one state.

> (or at least, once we know it's a version 5 tx)?

Just to be clear we add it to the `unknownKeyVals` when it is a version 4 transaction, we don't need to for a version 5.

Reply:
> Just to be clear we add it to the unknownKeyVals when it is a version 4 transaction, we don't need to for a version 5.

Oops, right, I had that backwards. Thanks for answering all my questions 👍